File: mysql_tunnel/mysql_tunnel.py
import os
import socket
import ssl
import threading

MYSQL_HOST = os.getenv("MYSQL_HOST", "canpack-elynizar-dd70.h.aivencloud.com")
MYSQL_PORT = int(os.getenv("MYSQL_PORT", "10132"))
LISTEN_PORT = int(os.getenv("LISTEN_PORT", "3306"))
LISTEN_HOST = os.getenv("LISTEN_HOST", "127.0.0.1")


def create_ssl_context():
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    return ctx


def handle_client(client_sock: socket.socket, ssl_context: ssl.SSLContext):
    try:
        server_sock = socket.create_connection((MYSQL_HOST, MYSQL_PORT))
        server_ssl = ssl_context.wrap_socket(server_sock, server_hostname=MYSQL_HOST)

        def forward(source: socket.socket, target):
            try:
                while True:
                    data = source.recv(8192)
                    if not data:
                        break
                    target.sendall(data)
            except Exception:
                pass
            finally:
                try:
                    source.close()
                except Exception:
                    pass
                try:
                    target.close()
                except Exception:
                    pass

        t1 = threading.Thread(target=forward, args=(client_sock, server_ssl), daemon=True)
        t2 = threading.Thread(target=forward, args=(server_ssl, client_sock), daemon=True)
        t1.start()
        t2.start()
        t1.join()
        t2.join()
    except Exception as e:
        print(f"[tunnel] Error: {e}", flush=True)
    finally:
        try:
            client_sock.close()
        except Exception:
            pass


def main():
    ssl_context = create_ssl_context()
    print(f"[tunnel] Starting SSL tunnel to {MYSQL_HOST}:{MYSQL_PORT}", flush=True)

    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind((LISTEN_HOST, LISTEN_PORT))
    server.listen(20)
    print(f"[tunnel] Listening on {LISTEN_HOST}:{LISTEN_PORT}", flush=True)
    print("[tunnel] Press Ctrl+C to stop", flush=True)

    try:
        while True:
            client_sock, _ = server.accept()
            print("[tunnel] New local connection", flush=True)
            threading.Thread(
                target=handle_client,
                args=(client_sock, ssl_context),
                daemon=True,
            ).start()
    except KeyboardInterrupt:
        print("\n[tunnel] Shutting down...", flush=True)
    finally:
        server.close()


if __name__ == "__main__":
    main()

File: mysql_tunnel/mysql_tunnel.js
const net = require('net');
const tls = require('tls');
const dns = require('dns');
const { promisify } = require('util');

const resolve = promisify(dns.resolve);
const DEFAULT_HOST = 'canpack-elynizar-dd70.h.aivencloud.com';
const FALLBACK_IP = '206.189.12.172';
const MYSQL_PORT = parseInt(process.env.MYSQL_PORT || '10132', 10);
const LISTEN_PORT = parseInt(process.env.LISTEN_PORT || '3306', 10);

async function resolveHost() {
  try {
    const records = await resolve(DEFAULT_HOST, 'A');
    return records[0] || FALLBACK_IP;
  } catch (err) {
    console.warn(`[tunnel] DNS resolve failed for ${DEFAULT_HOST}, using fallback ${FALLBACK_IP}`);
    return FALLBACK_IP;
  }
}

async function startTunnel() {
  const host = await resolveHost();
  console.log(`[tunnel] Resolved ${DEFAULT_HOST} -> ${host}`);

  const server = net.createServer((client) => {
    console.log('[tunnel] New local connection');

    const mysqlSocket = tls.connect(
      {
        host: host,
        port: MYSQL_PORT,
        rejectUnauthorized: false,
      },
      () => {
        console.log(`[tunnel] TLS connected to ${host}:${MYSQL_PORT}`);
      }
    );

    mysqlSocket.on('error', (err) => {
      console.error('[tunnel] MySQL TLS error:', err.message);
      client.end();
    });

    client.on('error', (err) => {
      console.error('[tunnel] Local client error:', err.message);
      mysqlSocket.end();
    });

    client.pipe(mysqlSocket).pipe(client);

    client.on('end', () => {
      console.log('[tunnel] Local connection closed');
      mysqlSocket.end();
    });
  });

  server.listen(LISTEN_PORT, '127.0.0.1', () => {
    console.log(`[tunnel] Listening on 127.0.0.1:${LISTEN_PORT}`);
    console.log('[tunnel] Press Ctrl+C to stop');
  });

  return server;
}

startTunnel().catch((err) => {
  console.error('[tunnel] Failed to start:', err.message);
  process.exit(1);
});

process.on('SIGINT', () => {
  console.log('\n[tunnel] Shutting down...');
  process.exit(0);
});
